### PR TITLE
:sparkles: Modules support 

### PR DESCRIPTION
**Pull request by Spiegie**
This is work in progress!

Firstly: Thank you for your Effort, that you did put into this project. I already love it.

Currently I'm trying to create a Module or Plugin-system for this Project. I want to try something new. Not like other IaC-Tools like saltstack or ansible.

The Idea is, that you can create binaries, which accept structured data. This Data can be written in toml, json, yaml or what ever you want. Even a Database or API is possible. Only the module has to understand it. You could just pass the data via host-vars. The Module (or Plugin) takes this Data and does certain Tasks like creating a database in mariadb or even installing and updating mariadb. The Modules could get rolled out by matching tags of the hosts.

This approach leads quite a bit away from the agentless nature of this project. I don't know a comparable IaC-Project. Maybe somebody else got ideas, how this Problem with this idea can be addressed, or if it even has to be addressed. 

But actually, I don't think, it has to be addressed. I would be very happy, to have a IaC-Tool, that tells a server once, how he should act and is adjustable via data. I like the Idea of a State, like Saltstack enforces it. But I also want to make the client (or host) less dependent on a master (except for the setup). In the words of Saltstack: I want to move the state to the Server and make the pillar even more flexible. Tricorder could act as management-tool for these Modules.

Currently, I'm uploading the binary to the Server via rsync. I'm sure, there is a better way. 

By the way, I'm relatively new to Rust. My code will get better :)

Please let me know what you think about this Idea. I'll respect your Opinion.

### CODE REVIEW

**Comment by linkdd** [comment]
Thank you for your feedback, it's greatly appreciated :heart:

Even though I did not have a lot of time for this project in the last year, it's not abandoned, I'd like to come back to it (maybe in the near future?).

Anyway, I like the idea! Uploading a binary to the remote and executing the command there is probably how I'd do it as well. A few remarks though:

 - I like to use https://gitmoji.dev for commit messages, if you can `git commit --amend` or `git rebase` to edit the message, it will be appreciated
 - the `ssh2` crate I use have support for uploading files through the SCP protocol, take a look at https://docs.rs/ssh2/latest/ssh2/#upload-a-file you might not need to use rsync
 - for error handling, instead of using `.unwrap()` or `.expect()`, I recommend to extend the [error type](https://github.com/linkdd/tricorder/blob/main/src/prelude/error.rs) and maybe implement the `From<T>` trait and then use the `?` operator
 - Within a task implementation (in `src/tasks/`), you should not use `println!` (stdout), but `eprintln!` (stderr), stdout should be used only for the task results (the JSON data, see the README for more information)

Again, thank you for your work, I'll be happy to review it and integrate it.

PS: I should finish the test suite, this is the one blocker for me to release as 1.0.0, but it can wait for your feature to land in 0.10 before 😄

**Comment by Spiegie** [comment]
Thank you for the answer and the remarks. 
I will consider all points and come back with a suitable commit.

**Comment by Spiegie** [comment]
I will create more documentation and some test, but I consider this complete for now. There are some things, that can be done better. (using str and lifetimes) but I would need much more time, to make it better. 

As soon as I'm done with the documentation, I'll remove the WIP from the MR-Title.

File: src/cli/module.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    if let Some(path) = arg {
+        Ok(String::from(path))
+    } else {
+        Err(Box::new(Error::MissingInput(
+            "No input file provided".to_string(),
+        )))
+    }
+}
+
+fn get_data_path(arg: Option<&str>) -> Option<String> {
```
Comment (linkdd) [inline src/cli/module.rs:29]: In this function, you can use `Option::map()`:

```rust
arg.map(String::from)
```

File: src/tasks/mod.rs [hunk 2]
```diff
@@ -4,3 +4,4 @@ pub mod exec;
 pub mod info;
 pub mod upload;
 pub mod download;
+pub mod module;
```
Comment (linkdd) [inline src/tasks/mod.rs:7]: Please add a newline at the end of file :)

File: src/tasks/module.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+impl GenericTask<String> for Task {
+    fn prepare(&self, host: Host) -> Result<String> {
+
+        let hostvars = host.vars.clone();
+        
+        let default = json!({});
+
+        
+        let host_var_data: &Value = hostvars.get(&format!("module_{}", self.module_name)).unwrap_or(&default);
+        println!("host_var_data: {:?}", host_var_data);
```
Comment (linkdd) [inline src/tasks/module.rs:46]: Don't forget to remove the debugging `println!` once you're done :)

**Review by linkdd — CHANGES_REQUESTED** [review]
It looks good so far.

I'll wait for your last changes on documentation and merge it.

Thank you 🙂

File: src/cli/module.rs (hunk 1 above)
Comment (Spiegie) [reply src/cli/module.rs:29]: cool :) thanks :)

File: src/tasks/module.rs (hunk 3 above)
Comment (Spiegie) [reply src/tasks/module.rs]: i hope they are all gone now

File: src/cli/module.rs (hunk 1 above)
Comment (Spiegie) [reply src/cli/module.rs:49]: done

File: src/tasks/mod.rs (hunk 2 above)
Comment (Spiegie) [reply src/tasks/mod.rs]: done

**Comment by Spiegie** [comment]
I want to include some tests, then I think I'm done

**Comment by linkdd** [comment]
Don't bother with the test suite, there is another PR for it ( #23 ) that I need to get back to.

Though, can you remove the TODO comments and open issues instead?

By the way, are you interested in being added to the members of this repository?

**Comment by Spiegie** [comment]
Ok thanks :) 
I think now you can fully review my code. I'm sure, here and there are some things, I could learn from you.

I would be honored to be a member of this repository. I'm planning to contribute more to this project.

**Pull request merged**